timer/config.go

```diff
@@ -0,0 +1,64 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package timer
+
+import (
+	"time"
+)
+
+// UnlimitedBudget is a special value for the budget that means the timer has no budget
+var UnlimitedBudget = ^time.Duration(0) - 1
+
+// InheritedBudget is a special value for the budget that means the timer should inherit the budget from its parent
+// It is the default value if no options such as WithBudget, WithUnlimitedBudget or WithInheritedBudget are provided
+var InheritedBudget = time.Duration(0)
+
+// Config is the configuration of a timer. It can be created through the use of options
+type Config struct {
+	// Budget is the time budget for the timer
+	Budget time.Duration
+	// Components store all the components of the timer
+	Components []string
+}
+
+func newConfig(options ...Option) Config {
+	config := Config{}
+	for _, option := range options {
+		option(&config)
+	}
+	return config
+}
```

Comment:
TBH this type is easy enough to form by hand without using the constructor + option functors... and the zero-value would be what you expect it to be... and all its fields are exported, too...

You could imagine using the `Config` as a template for timers (that you could store in a package-level variable), and then use `Config.Start()` to create a new started timer with the designated budget & components (or `.New()` to create a new un-started timer with the designated budget & components).

Reply:
Again I tried directly passing Config objects around but this did not end well and I had to modify 20 different tests to make things works again each time. Plus this is way easier for deprecation to works with the option system (that does not have any performance code as of right now)